### PR TITLE
[Seq2Seq Generation] Call encoder before expanding input_ids

### PR DESCRIPTION
**Pull request by sshleifer**
Proposing to call model.encoder before expanding `input_ids` to `effective_batch_size*num_beams`.

For Bart, this saves 1.5 GB of GPU mem on batch_size=6. Savings probably similar for T5 (untested).

Requires knowing which index of the encoder_outputs is associated with the batch dim (we need to expand this dimension), which is different between `Bart` and `T5`. This difference is encoded in the `self.encoder_outputs_batch_idx` variable.

This PR is WIP because `encoder_outputs_batch_idx` could be avoided if we transposed Bart's encoder_outputs, which I haven't tried.

### CODE REVIEW

**Comment by patrickvonplaten** [comment]
Like the change a lot! 
One question I asked myself: With this change the `encoder_outputs` which are the same point to the same memory address -> could that lead to problems? Probably not because the `encoder_outputs` are never changed, right?

File: src/transformers/modeling_bart.py [hunk 1]
```diff
@@ -113,6 +113,7 @@ class PretrainedBartModel(PreTrainedModel):
     config_class = BartConfig
     base_model_prefix = "model"
     pretrained_model_archive_map = BART_PRETRAINED_MODEL_ARCHIVE_MAP
+    encoder_outputs_batch_idx = 1  # outputs shaped (bs, ...)
```
Comment (patrickvonplaten) [inline src/transformers/modeling_bart.py]: would change the name to `encoder_outputs_batch_dim_idx`

File: src/transformers/modeling_t5.py [hunk 2]
```diff
@@ -457,6 +457,7 @@ class T5PreTrainedModel(PreTrainedModel):
     pretrained_model_archive_map = T5_PRETRAINED_MODEL_ARCHIVE_MAP
     load_tf_weights = load_tf_weights_in_t5
     base_model_prefix = "transformer"
+    encoder_outputs_batch_idx = 0  # outputs shaped (bs, ...)
```
Comment (patrickvonplaten) [inline src/transformers/modeling_t5.py]: would change the name to `encoder_outputs_batch_dim_idx`

File: src/transformers/modeling_utils.py [hunk 3]
```diff
@@ -929,6 +930,18 @@ def generate(
                 device=next(self.parameters()).device,
             )
             cur_len = 1
+            batch_idx = self.encoder_outputs_batch_idx
```
Comment (patrickvonplaten) [inline src/transformers/modeling_utils.py]: also here `batch_dim_idx`

File: src/transformers/modeling_utils.py [hunk 4]
```diff
@@ -929,6 +930,18 @@ def generate(
...
+            assert (
+                batch_size == encoder_outputs[0].shape[batch_idx]
+            ), f"expected encoder_outputs[0] to have 1st dimension bs={batch_size}, got {encoder_outputs[0].shape[1]} "
+            expanded_index = (
```
Comment (patrickvonplaten) [inline src/transformers/modeling_utils.py]: maybe also `expanded_idx` because we always use `idx` in the function?

**Review by thomwolf — APPROVED** [review]
LGTM.
I agree with the name changes proposed by @patrickvonplaten

**Pull request merged**